Public/Src/Tools/SandboxedProcessExecutor/Executor.cs

```diff
@@ -0,0 +1,322 @@
+                }
+                catch (BuildXLException ex)
+                {
+                    if (ex.LogEventErrorCode == NativeIOConstants.ErrorFileNotFound)
+                    {
+                        ConsoleError($"Failed to start process: '{info.FileName}' not found");
+                        return null;
+                    }
+                    else if (ex.LogEventErrorCode == NativeIOConstants.ErrorPartialCopy && (processRelaunchCount < ProcessRelauchCountMax))
+                    {
+                        ++processRelaunchCount;
+                        shouldRelaunchProcess = true;
+
+                        // Ensure that process terminates before relaunching it.
+                        if (process != null)
+                        {
+                            try
+                            {
+                                await process.GetResultAsync();
+                            }
+                            finally
+                            {
+                                process.Dispose();
+                            }
+                        }
+                    }
+                    else
+                    {
+                        ConsoleError($"Failed to start process '{info.FileName}': {ex.LogEventMessage} ({ex.LogEventErrorCode})");
+                        return null;
+                    }
+                }
+            }
+
+            return process;
+        }
+
+        private Stream TryOpenStandardInputStream(SandboxedProcessInfo info, out bool success)
+        {
+            success = true;
+
+            if (info.StandardInputSourceInfo == null)
+            {
+                return null;
+            }
+
+            try
+            {
+                if (info.StandardInputSourceInfo.Data != null)
+                {
+                    return new MemoryStream(CharUtilities.Utf8NoBomNoThrow.GetBytes(info.StandardInputSourceInfo.Data));
+                }
+                else  
+                {
+                    Contract.Assert(info.StandardInputSourceInfo.File != null);
+
+                    return FileUtilities.CreateAsyncFileStream(
+                        info.StandardInputSourceInfo.File,
+                        FileMode.Open,
+                        FileAccess.Read,
+                        FileShare.Read | FileShare.Delete);
+                }
+            }
+            catch (BuildXLException ex)
+            {
+                ConsoleError($"Unable to open standard input stream: {ex.ToString()}");
+                success = false;
+                return null;
+            }
+        }
+
+        private void ConsoleError(string errorMessage)
+        {
+            ConsoleColor original = Console.ForegroundColor;
+            Console.ForegroundColor = ConsoleColor.Red;
+            Console.Error.WriteLine(errorMessage);
+            Console.ForegroundColor = original;

```

Comment:
This needs some locking in case you have multiple threads writing to the console...